File: 문제해결기법/parallelogram/parallelogram.cpp
//
//  평행사변형.cpp
//  Algorithm
//
//  Created by 김동현 on 7/19/24.
//

#include <iostream>
#include <fstream>
#include <sstream>
#include <algorithm>
#include <cmath>
using namespace std;

struct Point
{
    int x, y;
};


inline bool compare(const Point &a, const Point &b)
{
    return a.x < b.x;
}

inline long long distance(const Point &a, const Point &b)
{
//    return sqrt(pow(b.x-a.x, 2) + pow(b.y-a.y, 2));
    return pow(b.x-a.x, 2) + pow(b.y-a.y, 2);
}


fstream inpfile("1.inp");
ofstream outfile("parallelogram.out");

Point point[4];
string line;

int main()
{
    // file I/O
    while (true)
    {
        getline(inpfile, line);
        istringstream iss(line);
        
        for (int i=0; i<4; i++)
        {
            iss >> point[i].x >> point[i].y;
        }
        
        // 기저사례
        if (point[0].x == 0 &&  point[0].y == 0 &&
            point[1].x == 0 &&  point[1].y == 0 &&
            point[2].x == 0 &&  point[2].y == 0 &&
            point[3].x == 0 &&  point[3].y == 0) break;
        
        // x좌표 기준 오름차순 정렬
        sort(point, point+4, compare);
        
        
        if (distance(point[0], point[1]) == distance(point[2], point[3]) && distance(point[0], point[2]) == distance(point[1], point[3]))
        {
            outfile << 1 << "\n";
        }
        else
        {
            outfile << 0 << "\n";
        }
    }
    inpfile.close();
    outfile.close();
    return 0;
}






//#include <iostream>
//#include <fstream>
//#include <sstream>
//using namespace std;
//
//struct Point
//{
//    int x, y;
//};
//
//fstream inpfile("1.inp");
//Point point[4];
//string line;
//
//int main()
//{
//    while (getline(inpfile, line))
//    {
//       istringstream iss(line);
//        for (int i=0; i<4; i++)
//        {
//            iss >> point[i].x >> point[i].y;
//            cout << point[i].x << " : " << point[i].y << "\n";
//        }
//    }
//    
//    return 0;
//}

File: 문제해결기법/평행사변형/parallelogram.swift
//
//  평행사변형.swift
//  Algorithm
//
//  Created by 김동현 on 7/19/24.
//

import Foundation


struct Point {
    var x: Int
    var y: Int
    
    static func compare(_ a: Point, _ b: Point) -> Bool {
        return a.x < b.x
    }

    static func distanceSquared(_ a: Point, _ b: Point) -> Int {
        return (b.x - a.x) * (b.x - a.x) + (b.y - a.y) * (b.y - a.y)
    }
}


@main
struct Main {
    static func main() {
        let inpfile = "1.inp"
        let outfile = "parallelogram.out"

        var points: [Point] = []

        do {
            let input = try String(contentsOfFile: inpfile, encoding: .utf8)
            let lines = input.components(separatedBy: .newlines)
            var output = ""

            for line in lines {
                if line.isEmpty { continue }
                let values = line.split(separator: " ").map { Int($0)! }
                for i in 0..<4 {
                    points.append(Point(x: values[2 * i], y: values[2 * i + 1]))
                }

                // 기저사례
                if points.allSatisfy({ $0.x == 0 && $0.y == 0 }) { break }

                // x좌표 기준 오름차순 정렬
                points.sort(by: Point.compare)

                let d1 = Point.distanceSquared(points[0], points[1])
                let d2 = Point.distanceSquared(points[2], points[3])
                let d3 = Point.distanceSquared(points[0], points[2])
                let d4 = Point.distanceSquared(points[1], points[3])

                if d1 == d2 && d3 == d4 {
                    output += "1\n"
                } else {
                    output += "0\n"
                }

                points.removeAll()
            }

            try output.write(toFile: outfile, atomically: true, encoding: .utf8)
        } catch {
            print("An error occurred: \(error)")
        }
    }
}
